main/src/ca/uwaterloo/flix/language/phase/jvm/GenReifiedSourceLocationClass.scala

```diff
@@ -0,0 +1,241 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.jvm
+
+import ca.uwaterloo.flix.api.Flix
+import org.objectweb.asm.{ClassWriter, Label}
+import org.objectweb.asm.Opcodes._
+
+object GenReifiedSourceLocationClass {
+
+  val sourceFieldName: String = "source"
+  val beginLineFieldName: String = "beginLine"
+  val beginColFieldName: String = "beginCol"
+  val endLineFieldName: String = "endLine"
+  val endColFieldName: String = "endCol"
+  val constructorDescriptor: String = AsmOps.getMethodDescriptor(List(JvmType.String, JvmType.PrimInt, JvmType.PrimInt, JvmType.PrimInt, JvmType.PrimInt), JvmType.Void)
+
+  def gen()(implicit flix: Flix): Map[JvmName, JvmClass] = {
+    val jvmName = JvmName.ReifiedSourceLocation
+    val bytecode = genByteCode(jvmName)
+    Map(jvmName -> JvmClass(jvmName, bytecode))
+  }
+
+  def genByteCode(name: JvmName)(implicit flix: Flix): Array[Byte] = {
+    // class writer
+    val visitor = AsmOps.mkClassWriter()
+
+    // internal name of super
+    val superClass = JvmName.Object
+
+    // Initialize the visitor to create a class.
+    visitor.visit(AsmOps.JavaVersion, ACC_PUBLIC + ACC_FINAL, name.toInternalName, null, superClass.toInternalName, null)
+
+    // Source of the class
+    visitor.visitSource(name.toInternalName, null)
+
+    def mkIntField(name: String): Unit = visitor.visitField(ACC_PUBLIC + ACC_FINAL, name, JvmType.PrimInt.toDescriptor, null, null).visitEnd()
+
+    visitor.visitField(ACC_PUBLIC + ACC_FINAL, sourceFieldName, JvmType.String.toDescriptor, null, null).visitEnd()
+    mkIntField(beginLineFieldName)
+    mkIntField(beginColFieldName)
+    mkIntField(endLineFieldName)
+    mkIntField(endColFieldName)
+
+    genConstructor(name, superClass, visitor)
+    genEqualsMethod(name, visitor)
+    genHashCode(name, visitor)
+    genToString(name, visitor)
+
+    visitor.visitEnd()
+    visitor.toByteArray
+  }
+
+  def genToString(name: JvmName, visitor: ClassWriter): Unit = {
+    val method = visitor.visitMethod(ACC_PUBLIC, "toString", AsmOps.getMethodDescriptor(Nil, JvmType.String), null, null)
+    method.visitCode()
+
+    method.visitTypeInsn(NEW, JvmName.StringBuilder.toInternalName)
+    method.visitInsn(DUP)
+    method.visitMethodInsn(INVOKESPECIAL, JvmName.StringBuilder.toInternalName, "<init>", AsmOps.getMethodDescriptor(Nil, JvmType.Void), false)
+    method.visitVarInsn(ALOAD, 0)
+    method.visitFieldInsn(GETFIELD, name.toInternalName, sourceFieldName, JvmName.String.toDescriptor)
+    method.visitMethodInsn(INVOKEVIRTUAL, JvmName.StringBuilder.toInternalName, "append", s"(${JvmName.String.toDescriptor})${JvmName.StringBuilder.toDescriptor}", false)
```

Comment:
@magnus-madsen Here is a lot of cases, using AsmOps.getMethodDescriptor,  where it could be nice to use JvmName/JvmType interchangably (things that only need .toDescriptor). Should I add a interface for it or is that too much effort considering it is going to be replaced?

Reply:
Its up to you. I assume eventually these classes with be replaced anyway!